data-plane/control-protocol/src/main/java/dev/knative/eventing/control/protocol/impl/TCPControlServerVerticle.java

```diff
@@ -0,0 +1,183 @@
+    this.port = port;
+    this.incomingMessageAddress = incomingMessageAddress;
+    this.outgoingMessageAddress = outgoingMessageAddress;
+
+    this.toAck = new HashMap<>();
+    this.enqueuedWaitingForConnection = new ArrayDeque<>();
+  }
+
+  @Override
+  public void start(Promise<Void> startPromise) throws Exception {
+    // Initialize event bus requester
+    this.serializedEventBusRequester =
+      new SerializedEventBusRequester(vertx.eventBus(), this.incomingMessageAddress, DELIVERY_OPTIONS);
+
+    // Start the event bus message listener
+    vertx.eventBus().<ControlMessage>localConsumer(this.outgoingMessageAddress)
+      .handler(this::handleOutboundMessage);
+
+    // Start the tcp server
+    this.server = vertx.createNetServer(createNetServerOptions());
+    server.connectHandler(this::newConnectionHandler);
+    server.exceptionHandler(t -> logger.error("Error in tcp server", t));
+    server
+      .listen()
+      .<Void>mapEmpty()
+      .onComplete(startPromise);
+  }
+
+  @Override
+  public void stop(Promise<Void> stopPromise) {
+    CompositeFuture.all(
+      (server != null) ? server.close() : Future.succeededFuture(),
+      (actualConnection != null) ? actualConnection.close() : Future.succeededFuture()
+    )
+      .<Void>mapEmpty()
+      .onComplete(stopPromise);
+  }
+
+  private NetServerOptions createNetServerOptions() {
+    return new NetServerOptions()
+      .setPort(port)
+      .setTcpKeepAlive(true);
+  }
+
+  private void newConnectionHandler(NetSocket netSocket) {
+    if (actualConnection != null) {
+      // Close the previous connection,
+      // the control protocol assumes only one connection is open at any given time
+      actualConnection.close();
+    }
+    actualConnection = netSocket;
+    actualConnection.handler(new ControlMessageParser(this::handleInboundMessage));
+    actualConnection.exceptionHandler(t -> logger.error("Error in connection", t));
+
+    // Dequeue the elements waiting for a connection
+    while (!enqueuedWaitingForConnection.isEmpty()) {
+      writeOnConnection(enqueuedWaitingForConnection.poll());
+    }
+  }
+
+  private void handleInboundMessage(ControlMessage message) {
+    if (message.opCode() == ControlMessage.ACK_OP_CODE) {
+      handleInboundAck(message);
+    } else {
+      emitMessageOnEventBus(message);
+    }
+  }
+
+  private void handleInboundAck(ControlMessage ackMessage) {
+    Message<ControlMessage> message = this.toAck.get(ackMessage.uuid());
+    if (message == null) {
+      logger.warn("Received an ack for an unknown uuid {}", ackMessage.uuid());
+    } else {
+      message.reply(ackMessage);
+    }
+  }
+
+  private void emitMessageOnEventBus(ControlMessage message) {
+    this.serializedEventBusRequester.request(message)
+      .onFailure(t -> logger
+        .error("Cannot route the incoming control message to {}: {}", this.incomingMessageAddress, message, t))
```

Comment:
We should revisit whether logging a message at this level is safe and we don't leak sensitive data, a TODO/issue?

Reply:
I can address it on this PR. I think the problem is that the `ControlMessageImpl#toString()` should not print the payload of the message, that's the only place where sensitive data can exist.